File: Data Structure/Ch02-Arrays/02-CreateOwnArray.java
/*
    Create own array
    class Array
    members: data, capacity, size
*/

public class Array{
    private int[] data;
    private int size;

    /*
        02 - Create own array
    */

    public Array(int capacity){
        data = new int[capacity];
        size = 0;
    }

    public Array(){
        this(10);
    }

    public int getSize(){
        return size;
    }

    public int getCapacity(){
        return data.length;
    }

    public boolean isEmpty(){
        return size==0;
    }

    /*
        03 - Add new element to array
    */

    //Add element to last
    public void addLast(int e){
        if(size == data.length){
            throw new IllegalArgumentException("AddLast failed, array is full");
        }
        data[size] = e;
        size++;

        //add(size, e);
    }

    //Add element to first
    public void addLast(int e){
        add(0, e);
    }

    //Add element to specific position (insert)
    public void add(int index, int e){
        if(size == data.length){
            throw new IllegalArgumentException("Add failed, array is full");
        }
        if(index < 0 || index > size){
            throw new IllegalArgumentException("Add failed, index");
        }
        for(int i=size-1; i>=index; i--){
            data[i+1] = data[i];
        }
        data[index] = e;
        size++;
    }

    /*
        04 - Query and Update
    */

    public int get(int index){
        if(index < 0 || index >= size){
            throw new IllegalArgumentException("Get failed, index");
        }
        return data[index];
    }

    public void set(int index, int e){
        if(index < 0 || index >= size){
            throw new IllegalArgumentException("Get failed, index");
        }
        data[index] = e;
    }

    @Override
    public String toString(){
        StringBuilder res = new StringBuilder();
        res.append(String.format("Array: Size = %d, capacity = %d\n", size, data.length));
        res.append('[');
        for(int i=0; i<size; i++){
            res.append(data[i]);
            if(i != size-1){
                res.append(", ");
            }
        }
        res.append(']');
    }

    /*
        05 - Contain, Find, Remove
    */
    public boolean contains(int e){
        for(int i=0; i<size; i++){
            if(data[i] == e){
                return true;
            }
        }
        return false;
    }

    public int find(int e){
        for(int i=0; i<size; i++){
            if(data[i] == e){
                return i;
            }
        }
        return -1;
    }

    public int remove(int index){
        if(index < 0 || index >= size){
            throw new IllegalArgumentException("Remove failed, index");
        }
        int ret = data[index];
        for(int i=index+1; i<size; i++){
            data[i-1] = data[i];
        }
        size--;
        return ret;
    }

    public int removeFirst(){
        return remove(0);
    }

    public int removeLast(){
        return remove(size-1);
    }

    //delete 1 e only
    public void removeElement(int e){
        int index = find(e);
        if(index != -1){
            remove(index);
        }
    }



}

File: Data Structure/Ch02-Arrays/02-CreateOwnArray.cpp
#include <iostream>
#include "02-CreateOwnArray.h"
using namespace std;

Array::Array(int s){
    data = new int[s];
    size = 0;
    length = s;
}

Array::~Array(){
    if(data){
        delete[] data;
    }
}

int& Array::operator[](int index){
    return data[index];
}

int Array::getSize(){
    return size;
}

int Array::getCapacity(){
    return length;
}

bool Array::isEmpty(){
    return size == 0;
}

void Array::add(int index, int e){
    if(index < 0 || index > size){
        cout << "error: index out of boundry" << endl;
        return;
    }
    if(size == length){
        cout << "error: array is full" << endl;
        return;
    }
    for(int i=size-1; i>=index; i--){
        data[i+1] = data[i];
    }
    data[index] = e;
    size++;
}

void Array::addLast(int e){
    add(size,e);
}

void Array::addFirst(int e){
    add(0,e);
}

int Array::get(int index){
    if(index < 0 || index >= size){
        cout << "error: index out of boundry" << endl;
        return -1;
    }
    return data[index];
}

void Array::set(int index, int e){
    if(index < 0 || index >= size){
        cout << "error: index out of boundry" << endl;
        return;
    }
    data[index] = e;
}

bool Array::contains(int e){
    for(int i=0; i<size; i++){
        if(data[i] == e){
            return true;
        }
    }
    return false;
}

int Array::find(int e){
    for(int i=0; i<size; i++){
        if(data[i] == e){
            return i;
        }
    }
    return -1;
}

int Array::remove(int index){
    if(index < 0 || index >= size){
        cout << "error: index out of boundry" << endl;
        return -1;
    }
    if(isEmpty()){
        cout << "error: array is empty" << endl;
        return -1;
    }
    int res = data[index];
    for(int i=index; i<size-1; i++){
        data[i] = data[i+1];
    }
    size--;
    return res;
}

int Array::removeFirst(){
    return remove(0);
}

int Array::removeLast(){
    return remove(size-1);
}

void Array::removeElement(int e){
    int index = find(e);
    if(index > -1){
        remove(index);
    }
}

